repo-client/src/main/kotlin/io/lionweb/repoclient/LionWebClient.kt

```diff
@@ -0,0 +1,734 @@
+            if (response.code == HttpURLConnection.HTTP_OK) {
+                val data =
+                    (response.body ?: throw IllegalStateException("Response without body when querying $url")).string()
+                debugFile("isNodeExisting-$nodeID.json") { data }
+                return processChunkResponse(data) { chunk ->
+                    val nodes = chunk.asJsonObject.get("nodes").asJsonArray
+                    !nodes.isEmpty
+                }
+            } else {
+                throw RuntimeException(
+                    "Something went wrong while querying $url: http code ${response.code}, body: ${response.body?.string()}",
+                )
+            }
+        }
+    }
+
+    fun getParentId(nodeID: String): String? {
+        require(nodeID.isNotBlank())
+        val body: RequestBody = "{\"ids\":[\"$nodeID\"] }".toRequestBody(JSON)
+        val url = "http://$hostname:$port/bulk/retrieve"
+        val urlBuilder = url.toHttpUrlOrNull()!!.newBuilder()
+        urlBuilder.addQueryParameter("depthLimit", "0")
+        val request: Request =
+            Request.Builder()
+                .url(urlBuilder.build())
+                .considerAuthenticationToken()
+                .post(body)
+                .build()
+        httpClient.newCall(request).execute().use { response ->
+            if (response.code == HttpURLConnection.HTTP_OK) {
+                val data =
+                    (response.body ?: throw IllegalStateException("Response without body when querying $url")).string()
+                debugFile("getParentId-$nodeID.json") { data }
+                return processChunkResponse(data) { chunk ->
+                    val nodes = chunk.asJsonObject.get("nodes").asJsonArray
+                    if (nodes.size() != 1) {
+                        throw UnexistingNodeException(
+                            nodeID,
+                            "When asking for the parent Id of $nodeID we were expecting to get one node back. " +
+                                "We got ${nodes.size()}",
+                        )
+                    }
+                    val node = nodes.get(0).asJsonObject
+                    require(nodeID == node.get("id").asString)
+                    val parentNode = node.get("parent")
+                    if (parentNode.isJsonNull) {
+                        null
+                    } else {
+                        parentNode.asString
+                    }
+                }
+            } else {
+                throw RuntimeException(
+                    "Something went wrong while querying $url: http code ${response.code}, body: ${response.body?.string()}",
+                )
+            }
+        }
+    }
+
+    fun storeTree(node: Node) {
+        treeStoringOperation(node, "store")
+    }
+
+    /**
+     * This operation is not atomic. We hope that no one is changing the parent at the very
+     * same time.
+     */
+    fun appendTree(
```

Comment:
The lionweb-repository does return a `version` attribute in each call, that could potentially be used as a sanity check.
However, this can only be used to check whether there was any change at all to the database between calls, it is not possible (at least with the current API) to know whether a specific node has changed.

Reply:
Yes, we should use it, especially when we do successive calls. I addes an issue for that (https://github.com/LionWeb-io/lionweb-java/issues/159)